src/server/es/aggs.js

```diff
@@ -502,43 +573,75 @@ export const textAggregation = async (
     aggsObj.aggs = updateAggObjectForMissingFields(nestedAggFields.missingFields, aggsObj.aggs);
   }
 
-  queryBody.aggs = {
-    [aggsName]: {
-      composite: {
-        sources: [
-          {
-            [field]: {
-              terms: {
-                field,
-                ...missingAlias,
-              },
+  // build up ES query if is nested aggregation
+  if (aggsNestedName) {
+    queryBody.aggs = {
+      [aggsNestedName]: {
+        nested: {
+          path: nestedPath,
+        },
+        aggs: {
+          [aggsName]: {
+            composite: {
+              sources: [
+                {
+                  [fieldNestedName]: {
+                    terms: {
+                      field: fieldNestedName,
+                      ...missingAlias,
+                    },
+                  },
+                },
+              ],
+              size: PAGE_SIZE,
             },
+            ...aggsObj,
           },
-        ],
-        size: PAGE_SIZE,
+        },
       },
-      ...aggsObj,
-    },
-  };
+    };
+  } else {
+    queryBody.aggs = {
+      [aggsName]: {
+        composite: {
+          sources: [
+            {
+              [field]: {
+                terms: {
+                  field,
+                  ...missingAlias,
+                },
+              },
+            },
+          ],
+          size: PAGE_SIZE,
+        },
+        ...aggsObj,
+      },
+    };
+  }
   let resultSize;
   let finalResults = [];
   /* eslint-disable */
   do {
+    // parse ES query result based on whether is doing nested aggregation or not (if `aggsNestedName` is defined)
     const result = await esInstance.query(esIndex, esType, queryBody); 
     resultSize = 0;
-        
-    result.aggregations[aggsName].buckets.forEach((item) => {
+
+    const resultBuckets = (aggsNestedName) ? result.aggregations[aggsNestedName][aggsName].buckets : result.aggregations[aggsName].buckets;
+
+    resultBuckets.forEach((item) => {
       const resultObj = processResultsForNestedAgg (nestedAggFields, item, {})
       finalResults.push({
-        key: item.key[field],
+        key: (fieldNestedName)? item.key[fieldNestedName] : item.key[field],
         count: item.doc_count,
         ...resultObj
       });
       resultSize += 1;
     });
-    const afterKey = result.aggregations[aggsName].after_key;
+    const afterKey = (aggsNestedName) ? result.aggregations[aggsNestedName][aggsName].after_key : result.aggregations[aggsName].after_key;
     if (typeof afterKey === 'undefined') break;
-    queryBody.aggs[aggsName].composite.after = afterKey;
+    (aggsNestedName) ? queryBody.aggs[aggsNestedName].aggs[aggsName].composite.after = afterKey : queryBody.aggs[aggsName].composite.after = afterKey;
   } while (resultSize === PAGE_SIZE);
```

Comment:
What's the behavior of this do/while loop? My understanding: It runs the query against the ES index and pushes each bucket of the response into `finalResults`. If the number of buckets in the result is equal to PAGE_SIZE, then it re-runs the same query against the ES index and pushes each bucket of the new response into `finalResults`? So I guess it relies on the instance returning paginated results for repeated queries?

Reply:
Yes you are right! So if the `result size == PAGE_SIZE` and there is a `after_key` then we just update the query and repeat to get paginated results
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-composite-aggregation.html#_pagination

Reply:
Gotchu, I hadn't noticed the afterKey part. Thanks!